File: sdk/python/express_relay/searcher/examples/testing_searcher_svm.py
import asyncio
import random

from express_relay.models import Opportunity
from express_relay.svm.limo_client import OrderStateAndAddress
from solders.keypair import Keypair

from .simple_searcher_svm import SimpleSearcherSvm, get_parser


class TestingSearcherSvm(SimpleSearcherSvm):
    def __init__(
        self,
        server_url: str,
        private_key: Keypair,
        bid_amount: int,
        chain_id: str,
        svm_rpc_endpoint: str,
        fill_rate: int,
        with_latency: bool,
        bid_margin: int,
        api_key: str | None = None,
    ):
        super().__init__(
            server_url, private_key, bid_amount, chain_id, svm_rpc_endpoint, api_key
        )
        self.fill_rate = fill_rate
        self.with_latency = with_latency
        self.bid_margin = bid_margin

    async def opportunity_callback(self, opp: Opportunity):
        if self.with_latency:
            latency = 0.5 * random.random()
            self.logger.info(f"Adding latency of {latency * 100}ms")
            await asyncio.sleep(latency)
        return await super().opportunity_callback(opp)

    async def get_bid_amount(self, order: OrderStateAndAddress) -> int:
        return self.bid_amount + random.randint(-self.bid_margin, self.bid_margin)

    def get_input_amount(self, order: OrderStateAndAddress) -> int:
        return min(
            super().get_input_amount(order),
            order["state"].initial_input_amount * self.fill_rate // 100,
        )


async def main():
    parser = get_parser()
    parser.add_argument(
        "--fill-rate",
        type=int,
        default=100,
        help="How much of the initial order size to fill in percentage. Default is 100%",
    )
    parser.add_argument(
        "--with-latency",
        required=False,
        default=False,
        action="store_true",
        help="Whether to add random latency to the bid submission. Default is false",
    )
    parser.add_argument(
        "--bid-margin",
        required=False,
        type=int,
        default=0,
        help="The margin to add or subtract from the bid. For example, 1 means the bid range is [bid - 1, bid + 1]. Default is 0",
    )
    args = parser.parse_args()

    if args.private_key:
        searcher_keypair = Keypair.from_base58_string(args.private_key)
    else:
        with open(args.private_key_json_file, "r") as private_key_file:
            searcher_keypair = Keypair.from_json(private_key_file.read())

    searcher = TestingSearcherSvm(
        args.endpoint_express_relay,
        searcher_keypair,
        args.bid,
        args.chain_id,
        args.endpoint_svm,
        args.fill_rate,
        args.with_latency,
        args.bid_margin,
        args.api_key,
    )

    await searcher.client.subscribe_chains([args.chain_id])

    task = await searcher.client.get_ws_loop()
    await task


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/js/src/examples/testingSearcherSvm.ts
import { Opportunity } from "../index";

import * as anchor from "@coral-xyz/anchor";
import { Keypair } from "@solana/web3.js";
import { OrderStateAndAddress } from "@kamino-finance/limo-sdk/dist/utils";
import { getKeypair, makeParser, SimpleSearcherSvm } from "./simpleSearcherSvm";

class SearcherLimo extends SimpleSearcherSvm {
  private readonly fillRate: anchor.BN;

  constructor(
    endpointExpressRelay: string,
    chainId: string,
    searcher: Keypair,
    endpointSvm: string,
    bid: number,
    fillRate: number,
    public withLatency: boolean,
    public bidMargin: number,
    public apiKey?: string,
  ) {
    super(endpointExpressRelay, chainId, searcher, endpointSvm, bid, apiKey);
    this.fillRate = new anchor.BN(fillRate);
  }

  async getBidAmount(): Promise<anchor.BN> {
    const margin = new anchor.BN(
      Math.floor(Math.random() * (this.bidMargin * 2 + 1)) - this.bidMargin,
    );
    return this.bid.add(margin);
  }

  async opportunityHandler(opportunity: Opportunity): Promise<void> {
    if (this.withLatency) {
      const latency = Math.floor(Math.random() * 300);
      console.log(`Adding latency of ${latency}ms`);
      await new Promise((resolve) => setTimeout(resolve, latency));
    }
    return super.opportunityHandler(opportunity);
  }

  protected getInputAmount(order: OrderStateAndAddress): anchor.BN {
    return anchor.BN.min(
      super.getInputAmount(order),
      order.state.initialInputAmount.mul(this.fillRate).div(new anchor.BN(100)),
    );
  }
}

async function run() {
  const argv = makeParser()
    .option("fill-rate", {
      description:
        "How much of the initial order size to fill in percentage. Default is 100%",
      type: "number",
      default: 100,
    })
    .option("with-latency", {
      description:
        "Whether to add random latency to the bid submission. Default is false",
      type: "boolean",
      default: false,
    })
    .option("bid-margin", {
      description:
        "The margin to add or subtract from the bid. For example, 1 means the bid range is [bid - 1, bid + 1]. Default is 0",
      type: "number",
      default: 0,
    })
    .parseSync();
  const searcherKeyPair = getKeypair(argv.privateKey, argv.privateKeyJsonFile);
  const simpleSearcher = new SearcherLimo(
    argv.endpointExpressRelay,
    argv.chainId,
    searcherKeyPair,
    argv.endpointSvm,
    argv.bid,
    argv.fillRate,
    argv.withLatency,
    argv.bidMargin,
    argv.apiKey,
  );
  await simpleSearcher.start();
}

if (require.main === module) {
  run();
}
